src/Model/Operation/ProductSyncHandler.php

```diff
@@ -178,35 +201,55 @@ private function doUpsertOperation(
         $operation->upsert();
     }
 
-    private function processProductVariants(SalesChannelProductEntity $product, SalesChannelContext $context): array
-    {
+    private function processProductVariants(
+        ProductEntity $product,
+        SalesChannelContext $context,
+        Account $account,
+        array $ids,
+    ): ProductCollection {
         $variantConfig = $product->getVariantListingConfig();
         $configuratorGroups = array_filter(
             $variantConfig?->getConfiguratorGroupConfig() ?? [],
             static fn (array $config) => $config['expressionForListings'],
         );
 
         if (!$product->getChildCount() || !($variantConfig instanceof VariantListingConfig)) {
-            return [$product];
+            return new ProductCollection([$product]);
         }
 
         $mainProducts = [];
```

Comment:
Please initialize it already as collection.

```suggestion
        $mainProducts = new ProductCollection();
```

Reply:
Should I change the return type of the methods `handleCheapestVariant`, `handleMainVariant`, `handleConfiguratorGroups`, `handleFirstActiveVariant` to `ProductCollection`?

Reply:
The `handleConfiguratorGroups` method returns an array. In order to assign the elements of the array to a `ProductCollection`, I would need to iterate through it and add each element individually using the add method. I would only modify that method to avoid using `foreach`.

Reply:
Yes, please also change the other return types

You can use the merge function to merge 2 collections.

Reply:
Sorry, I forgot that some of the functions only return one product. Please only one more change 😅:

`handleConfiguratorGroups` is fine with returning a collection, as we are expecting more products. Please change the other handle functions back to `?ProductEntity` + the `add()` within `processProductVariants`, as we are only expecting one product or `null`.

Reply:
I've made the necessary changes.